### ISSUE TITLE
`odo dev`: Trigger sync and push from the outside

### ISSUE THREAD

**Issue by cdrage** [redhat-developer/odo#5986]
/kind user-story

## User Story
- As a user or an IDE plugin developer,
- I want to be able to trigger sync and push "from the outside" (outside of the "odo dev" session, from a separate process)
- So that I can control those phases and trigger push only after I made multiple edits because I don't need to see every single small change and on a complex application it can be expensive to trigger rebuild

## Acceptance Criteria 
- [ ] `odo dev` should expose a side server that exposes an API (endpoints to discuss) that could be used to trigger sync and push
- [ ] Server should be started regardless if odo dev is running with or without `--no-watch`
- [ ] Server should be open for later extension. Might be used later to expose more information and control other things
- [ ] Information about the API (like its port) should be stored somewhere (maybe in devstate.json file)

## Notes
- This is part 2 of #5634: ❓❗ there should be a way to trigger sync "from the outside" (for example by sending signal to `odo dev` process (what about Windows?), or running new odo command in another terminal for example `odo dev --push`)
- For the record, we mentioned the ability for odo dev to run a side server that exposes an API that could be used by external tools to control its behavior (like triggering sync) or retrieve additional information.

This is a nice-to-have for v3, but not mandatory for GA of v3.

## Links 
- Related user story: https://github.com/redhat-developer/odo/issues/5634

### Child issues:
- [x] #6301
- [x] #6302

/kind user-story

**Comment by valaparthvi** [comment]
`odo dev` adds an async watcher/go routine/server that watches for a signal to manually trigger.
The signal will be an HTTP request.

Any file changes will send the HTTP request and that will trigger the sync.
Triggers for sync:
- manual trigger with keypress
- file change detection
- HTTP request

**Comment by valaparthvi** [comment]
The IDE plugins support keypress and there is no need to send a request.
@kadel @mohitsuman Can you confirm if this issue is still required?

**Comment by mohitsuman** [comment]
If there is an odo command to do the push, it is much better to integrate rather than passing a keypress to the terminal from IDE. And we can also handle errors if that command fails or it needs to show some specific message.

**Comment by mohitsuman** [comment]
@valaparthvi We tried using the keypress method and it does not work as expected. I would recommend having a specific odo command to sync and push changes.

**Comment by kadel** [comment]
> @valaparthvi We tried using the keypress method and it does not work as expected. I would recommend having a specific odo command to sync and push changes.

Even if we decide to implement the specific command, we will first need a way to communicate between two odo processes. We will do this by implementing a simple HTTP API on `odo dev` side.

I'm not sure about the dedicated command for this TBH. Once we implement this simple API it will be better for plugins to use that. It is going to be much easier and more stable than dealing with processes. In general, invoking HTTP requests is easier than dealing with system processes.

**Comment by kadel** [comment]
Here is my first draft of how the HTTP API could look like.

### GET /instance
Get information about the this 'odo dev' instance.

return data example
```json
{
  "componentDirectory": "/Users/user/Documents/myproject",
  "pid": 42
}
```

### DELETE /instance
Stop this 'odo dev' instance. The same as 'ctrl+c' on terminal

return data example
``` json
{
  "message": "'odo dev' instance with pid: 42 is shuting down."
}
```

### GET /component
Information about the component. This should return the same data as `odo describe component -o json`



### POST /component/command
Instruct 'odo dev' to perform the given command on the component

This expects the name of the command to be passed with the request.

request data:
```json
{
  "name": "push"
}
```

return data example

```json
{
  "message": "push was successfully executed"
}
```




It might look more complex than it needs to be, but I wanted to ensure that we can easily extend it in the future without breaking backward compatibility.

I've created OpenAPI we can use it to have automatically generated documentation, and consumers of this API can use it to create mock servers for development testing. 


OpenAPI spec: https://ododev.docs.apiary.io/api-description-document
Rendered Documentation: https://ododev.docs.apiary.io/

There is also an automatically generated mock server at `https://private-anon-13f9ca22d0-ododev.apiary-mock.com`

```
curl https://private-anon-13f9ca22d0-ododev.apiary-mock.com/instance
curl -X DELETE https://private-anon-13f9ca22d0-ododev.apiary-mock.com/instance
curl -X POST -d '{"name": "push"}'  https://private-anon-13f9ca22d0-ododev.apiary-mock.com/component/command
curl https://private-anon-13f9ca22d0-ododev.apiary-mock.com/component
```

**Comment by rm3l** [comment]
/close

All issues completed.
